calcom/cal.com
# feat: Populate gCal calendar cache via webhooks

fixes #11524

## What does this PR do?

- [NEW Loom Demo](https://www.loom.com/share/a3e807f13c324948b7b3ce195a1f508e?t=1)
- [OLD Loom Demo](https://www.loom.com/share/8ff219dc1c88434195306e826efd74e3)
- [Inbound Webhooks Cache Refresh Demo](https://www.loom.com/share/c8ef973daabf4cc3ae8c95427265b002?sid=c094995c-143e-4736-9354-a08c083b728b)
- Adds a mechanism to listen to changes in Google Calendar
- Adds an inbound webhook to pre-populate gCal cache
- Adds a team feature flag so we can enable this progressively
- Adds a cron job that handles calendar subscriptions and renewals before expiration.

### Calendar Cache and Cron Jobs Enhancements:
* [`apps/web/app/api/calendar-cache/cron/route.ts`](diffhunk://#diff-c5745b8ea9b8e396b95f3bf53d12dd572c856524987e667ee04c8c3ecc7ddeb0R1-R7): Exported the `GET` handler and set cache revalidation to zero to ensure fresh data each minute.
* [`apps/web/cron-tester.ts`](diffhunk://#diff-dd20add676725b08806c51a7813907a1cfe4f07eeaf5b5b38b9589d6e19e0ddeR2-R9): Added dotenv configuration for environment variables and updated cron job endpoints to use the correct API key parameter. [[1]](diffhunk://#diff-dd20add676725b08806c51a7813907a1cfe4f07eeaf5b5b38b9589d6e19e0ddeR2-R9) [[2]](diffhunk://#diff-dd20add676725b08806c51a7813907a1cfe4f07eeaf5b5b38b9589d6e19e0ddeL23-R26)
* [`apps/web/vercel.json`](diffhunk://#diff-361f945cebcf11354b7400e3c2b5f31c66b73f5c3a5b3f428a0a0f78d3bc898bR11-R14): Added a new cron job entry for the calendar cache API.

### API Handler Refactoring:
* [`apps/web/pages/api/availability/calendar.ts`](diffhunk://#diff-af4ba7e365d5576f6c3ef01accffaabd2040cf0767698693d6eb9816403b2241R6-R25): Refactored the API handler by splitting it into separate GET, POST, and DELETE handlers, and introduced shared authentication middleware. [[1]](diffhunk://#diff-af4ba7e365d5576f6c3ef01accffaabd2040cf0767698693d6eb9816403b2241R6-R25) [[2]](diffhunk://#diff-af4ba7e365d5576f6c3ef01accffaabd2040cf0767698693d6eb9816403b2241L38-R71) [[3]](diffhunk://#diff-af4ba7e365d5576f6c3ef01accffaabd2040cf0767698693d6eb9816403b2241L77-R87) [[4]](diffhunk://#diff-af4ba7e365d5576f6c3ef01accffaabd2040cf0767698693d6eb9816403b2241L89-R114)
* [`packages/app-store/googlecalendar/api/webhook.ts`](diffhunk://#diff-9e25400e22ffe57e9552ba9afb1fe3a7fee24c56de5ca32ade21839135adc53eR1-R47): Added a new webhook handler to validate requests and update calendar availability cache.

### Authentication and Security Improvements:
* [`apps/web/pages/api/cron/calendar-cache-cleanup.ts`](diffhunk://#diff-4ccfd89bc338d40763cb78faab2bf93b3996228ab9fa712e866c73d23d34ed1bR6-R11): Added API key validation to ensure only authorized requests can access the cron job endpoint.

### Calendar Cache and Google Calendar Integration:
* [`packages/app-store/googlecalendar/lib/CalendarService.ts`](diffhunk://#diff-6bfbaf63c841dc9343dba0c7487e30a546ffe2a63c9d9418379d0dbb88644d27R6-R11): Integrated the `CalendarCache` class for handling calendar cache operations and removed outdated cache handling functions. [[1]](diffhunk://#diff-6bfbaf63c841dc9343dba0c7487e30a546ffe2a63c9d9418379d0dbb88644d27R6-R11) [[2]](diffhunk://#diff-6bfbaf63c841dc9343dba0c7487e30a546ffe2a63c9d9418379d0dbb88644d27R24) [[3]](diffhunk://#diff-6bfbaf63c841dc9343dba0c7487e30a546ffe2a63c9d9418379d0dbb88644d27R45-R53)
* [`packages/app-store/googlecalendar/lib/CalendarService.test.ts`](diffhunk://#diff-70c6b1aad6ea84b27dffaeeb5b9e538476bca34a4d655f9e726be7142a8c2336L8-R10): Updated tests to cover new calendar cache functionalities, including cache hits, misses, and calendar watch/unwatch operations. [[1]](diffhunk://#diff-70c6b1aad6ea84b27dffaeeb5b9e538476bca34a4d655f9e726be7142a8c2336L8-R10) [[2]](diffhunk://#diff-70c6b1aad6ea84b27dffaeeb5b9e538476bca34a4d655f9e726be7142a8c2336L35-R37) [[3]](diffhunk://#diff-70c6b1aad6ea84b27dffaeeb5b9e538476bca34a4d655f9e726be7142a8c2336L45-R62) [[4]](diffhunk://#diff-70c6b1aad6ea84b27dffaeeb5b9e538476bca34a4d655f9e726be7142a8c2336L70-R75) [[5]](diffhunk://#diff-70c6b1aad6ea84b27dffaeeb5b9e538476bca34a4d655f9e726be7142a8c2336R88-R154)

These changes aim to improve the efficiency, security, and reliability of calendar-related features in the application.



## TODO (Follow up maybe)

- [x] Add a mechanism to auto-refresh about-to-expire webhooks
- [ ] Add a mechanism to bypass cache either via request header or query parameter
- [ ] Add a way to bulk enable for ALL users, once we feel confident enough to do so

## Type of change




## How should this be tested?

1. Create an accessible tunnel from a service like tmole
1. Put the resulting URL `.env` like `GOOGLE_WEBHOOK_URL="https://xxxxxxxxx.tunnelmole.net"`
1. Make sure your running the cron using `yarn dev:cron` inside apps/web directory. 
1. As a team member, Install a Google Calendar from the App Store
1. Once installed, enable a selected calendar
1. In the DB create a new `TeamFeature` entry with the proper `teamId` and a `featureId` named `calendar-cache`
1. You should receive a POST request populating your first cache
1. Add an event to your Google Calendar, you should receive another POST request, refreshing the cache.


## Mandatory Tasks
